clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -166,7 +167,10 @@ public synchronized void update(Cluster cluster, long now) {
             listener.onMetadataUpdate(cluster);
 
         // Do this after notifying listeners as subscribed topics' list can be changed by listeners
-        this.cluster = this.needMetadataForAllTopics ? getClusterForCurrentTopics(cluster) : cluster;
+        final Set<String> topicsCopy = topics();
```

Comment:
Is this really needed? Aren't all modifications of `topics` done within a `synchronized` block?